### ISSUE TITLE
[Roadmap] vLLM Roadmap Q3 2025

### ISSUE THREAD

**Issue by simon-mo** [vllm-project/vllm#20336]
This page is accessible via [roadmap.vllm.ai](https://roadmap.vllm.ai/)

This is a living document! For each item here, we intend to link the RFC as well as discussion Slack channel in the [vLLM Slack](https://slack.vllm.ai)

---

## Core Themes

*In Q3, we continue to iterate towards vLLM 1.0 by fully removing the V0 code path, optimizing and extending the core scheduler, making sure vLLM can serve the most demanding workloads of the world, and enhancing the out of box usability and performance.* 

### V1 Engine

* [x] V0 Features Parity and Native Features (\#sig-v1)  
  * [x] Pooling Model (#16188)
  * [x] Mamba Model (https://github.com/vllm-project/vllm/pull/19327)
  * [x] Priority Scheduling (https://github.com/vllm-project/vllm/pull/19057)
  * [x] Custom Logits Processing (#17799)
  * [x] CPU KV Cache (#19854)
  * [x] Investigate Encoder-Decoder Support (https://github.com/vllm-project/vllm/pull/21088)
* [ ] Performance   
  * [x] Async Scheduling https://github.com/vllm-project/vllm/pull/19970
  * [ ] Optimize Input Preparation (Persistent Batch V2)  
  * [ ] Speculative Decoding Enhancements (Suffix Decoding, CUDA Graph/torch.compile support)  
  * [ ] Multimodal Processing  
    * [ ] Simplification  
    * [x] Parallel Input Processing (#17546)
    * [ ] Reduce Serialization & Broadcasting Overheads (#19702)
    * [ ] Investigate streaming input and output  
* [ ] Design Documentation  
  * [x] Hybrid Memory Allocator  
  * [ ] Core Scheduler Design  
  * [ ] Speculative Decoding Design  
  * [x] Hardware Platform Guide  
  * [x] Model API Guide

### User Experience

* [ ] Fit and Finish (\#feat-startup-ux)
  * [x] Fast startup (https://github.com/vllm-project/vllm/issues/19824)
    * [x] Validate @config in pre-commit instead of dynamically (https://github.com/vllm-project/vllm/pull/20200)
    * [x] Skip GC during cudagraph capture (https://github.com/vllm-project/vllm/pull/21146)
    * [x] Skip p2p check by default (https://github.com/vllm-project/vllm/pull/21904)
    * [x] Optimize torch.compile cold-start (https://github.com/vllm-project/vllm/pull/20790
    * [ ] Optimize torch.compile warm-start
  * [x] Clean startup log  
  * [x] Clean up configuration items  
  * [x] Performance Tuning Guide (https://github.com/vllm-project/recipes)
  * [ ] Accelerator UX Audit and Document Feature Coverage  
* [ ] Stability and Testing  
  * [ ] Comprehensive Reproducible Performance Suite    
  * [ ] Enhance and Report Accuracy Suite  
  * [x] Large Scale Deployment Tested in CI  
  * [ ] Stress and Longevity Testing  
  * [ ] Improve the Stability of the vLLM-torch.compile Integration.  
  * [ ] Robust Tool Use Parsing
* [ ] Operational Experience  
  * [ ] Request Level SLO Targeting & Enhanced Autoscaling/Tuning  
  * [ ] Improve Logging and Tracing Code Path  
  * [x] Debugging tool for perf profiling and numerics

### Large Scale Serving

* [ ] Stable Scale-Out Serving for Mixture-of-Experts Models  
  * [x] Enhance and Document Data Parallelism (#20768)
  * [x] Stabilize Expert Parallelism Routing x GEMM Options   
  * [x] Expert Parallel Load Balancing (https://github.com/vllm-project/vllm/pull/18343)
  * [ ] Transfer KV Cache through CPU  
  * [x] Communication & Computation Overlap  
* [ ] Disaggregated Serving   
  * [ ] Standardized Dataflow between P \<\> D  
  * [ ] Autoscaling P & D Replicas  
  * [ ] Multi-modality Support (#20799)
  * [ ] Speculative Decoding Support  
  * [ ] Prefill-Only Mode  
* [ ] Elastic EP and Fault Tolerance  
* [ ] Enhancement to the KV Transfer API for Request Migration and KV Cache Priming  
* [x] Investigate Context Parallelism

### Models

* [x] Support Multiple Training and Model Authoring Frameworks by Opening up the Interface for Tokenizer, Configuration, and Processor.   
* [x] Investigate Sparse Attention Mechanism   
* [x] Performance Enhancements for Small Models (\<1B scale)

### Hardwares

* [ ] NVIDIA  
  * [x] Enhance Blackwell Support
    * [x] FP4 GEMM, MoE
    * [x] Optimized Attention (FlashInfer)
    * [x] CUTLASS MLA (https://github.com/vllm-project/vllm/pull/17625)
  * [ ] GB200 NVL72   
* [ ] AMD  
  * [ ] MI350X: MXFP4 Support  
  * [ ] Large Scale Serving Support  
  * [ ] Official Wheels and Container Distributed  
  * [ ] Document Feature Party and Performance Numbers  
* [ ] TPU  
  * [x] Progress in Ironwood Support  
  * [ ] Official Wheels and Container Distributed  
  * [ ] Document Feature Party and Performance Numbers  
* [ ] Neuron  
  * [x] Plugin for V1 Architecture  
  * [ ] Document Feature Party and Performance Numbers  
* [ ] Intel  
  * [x] Stable CPU Release with Wheels and Containers  
  * [x] Stable XPU Support   
  * [x] HPU (Gaudi) Move to Plugin  
* Platform Plugins  
  * [ ] Stable and Tested Interfaces
  * [x] vLLM Ascend: https://github.com/vllm-project/vllm-ascend/issues/1168

### Use Cases

* [ ] RLHF  
  * [x] Test Popular Framework that Integrate with vLLM for Performance and Prevent Breakages   
  * [ ] Weight loading optimization for syncing and resharding  
  * [ ] Custom checkpoint loader, custom model format  
  * [ ] Multi-turn scheduling  
* [ ] Evaluation  
  * [ ] Support Full Determinism (with/without prefix cache) Regardless of Batching Order  
* [ ] Batch Inference  
  * [ ] Simple Data Parallel Router for Scale Out with Prefix Caching  
  * [x] CPU KV cache offloading  
* [ ] Explore Bundling of Configuration for Specializations  
  * [ ] Low Latency Code Completion  
  * [ ] High Throughput Multi-turn Agentic Rollout  
  * [ ] Large Scale Image and Video Understanding   
  * [ ] Transformer Based Item Recommendation

### FAQ

*When will vLLM release 1.0?*  
We believe 1.0 means API stability and great user experience. We will not commit a date for the exact release. The criteria for 1.0 are: 

* Stable user facing API such as CLI, LLM, AsyncLLMEngine .  
* Stable developer API such as logits processor, kv connector, model interface, hardware/platform plugin interfaces  
* Polished out-of-box user experience.
---

If any of the items you wanted is not on the roadmap, your suggestion and contribution is strongly welcomed! Please feel free to comment in this thread, open feature request, or create an RFC.

Historical Roadmap: #15735, #11862, #9006, #5805, #3861, #2681, #244

**Comment by robertgshaw2-redhat** [comment]
I think we should create a sig and tracking RFC for each major item above

**Comment by skylee-01** [comment]
> I think we should create a sig and tracking RFC for each major item above

This helps to see the progress of each feature and assists contributors in deciding whether they can participate in this project.

**Comment by dtransposed** [comment]
What does "Performance Enhancements for Small Models (<1B scale)" mean?

**Comment by robertgshaw2-redhat** [comment]
I would like to add a high level goal of refactoring, improving performance, and stabilizing the tool calling implementation in vllm.

**Comment by rexlee8776** [comment]
@youkaichao Since DeepSeek announced in April to collaborate with upstream communities and contribute back to the open source community. We know that vLLM is one of these upstream communities and it's great to see the collaboration between great opensource projects.
I'd like to ask if DeepSeek's related features or optimizations be included in vLLM Roadmap Q3?

- [OpenSourcing_DeepSeek_Inference_Engine](https://github.com/deepseek-ai/open-infra-index/tree/main/OpenSourcing_DeepSeek_Inference_Engine)

**Comment by foreverlms** [comment]
Like `flashinfer`, `transformer engine` could also be the backend of attention. Will `integration of transformer engine` in the road map in near future?

**Comment by ECMGit** [comment]
Question regarding [Flash Attention](https://github.com/Dao-AILab/flash-attention) and [vLLM Flash Attention](https://github.com/vllm-project/flash-attention), I know so far this two flash attention is used in parallel when you set `VLLM_ATTENTION_BACKEND=FLASH_ATTN`, sometimes it cause problems. Are we going to plan to sticky one of them or still make them two used both in vllm?

**Comment by Isaka-code** [comment]
Are there any recommended settings or best practices for using vLLM with the RTX 5090? I'm having trouble getting it to run properly.

**Comment by celsowm** [comment]
> Are there any recommended settings or best practices for using vLLM with the RTX 5090? I'm having trouble getting it to run properly.

Docker lastest version is fine for me now on blackwell

**Comment by boposki** [comment]
Can current CONNECTOR be used for CPU KV cache offloading？

**Comment by hsliuustc** [comment]
what do you mean by Simple Data Parallel Router for Scale Out with Prefix Caching? what's the usage case for this feature?

**Comment by Oldpan** [comment]
I'm curious about the TODO for "Performance Enhancements for Small Models (<1B scale)" since my scenario is very similar. I'd like to explore if I can contribute

**Comment by cd-yang** [comment]
Currently, `AsyncLLMEngine` is a powerful tool, and I'm finding it incredibly useful in offline inference. However, to further enhance its capabilities and align it with typical online inference scenarios, I'd greatly appreciate the addition feature:

* **Chat Interface:** Integrating a dedicated chat interface would significantly streamline the development of conversational AI applications. 

Thank you for considering this request!

**Comment by yangqinghao-cmss** [comment]
Will `enable_prompt_embeds` be supported in V 1?

**Comment by BobDu** [comment]
Is it possible to add this feature to the roadmap?

Add -d/--detach option for `vllm serve`
#18065

**Comment by newball22** [comment]
I’m trying to contribute to this simple issue. Could you please tell me how to proceed with the integration?
<img width="775" height="598" alt="Image" src="https://github.com/user-attachments/assets/594421a2-b51b-45fa-b422-c08f4a7f08a9" />

**Comment by michael604work** [comment]
Is multi-GPU inference setting, TP in particular, supported for MI350X?

Thank you so much :-) :-)

**Comment by celsowm** [comment]
Please add support to structured output when using async schedule

**Comment by roxroxroxrox** [comment]
> I think we should create a sig and tracking RFC for each major item above

too many things are out of control.  RFC is suitable for corp management
